src/client/Microsoft.Identity.Client/Http/HttpManager.cs

```diff
@@ -76,6 +77,78 @@ protected virtual HttpClient GetHttpClient()
             return await ExecuteWithRetryAsync(uri, headers, body, HttpMethod.Post, logger, doNotThrow: true).ConfigureAwait(false);
         }
 
+        public async Task<HttpResponse> SendGetWithTimeoutAsync(
+            Uri endpoint,
+            IDictionary<string, string> headers,
+            ICoreLogger logger,
+            int timeoutInMs)
+        {
+            CancellationTokenSource cts = new CancellationTokenSource();
+            cts.CancelAfter(timeoutInMs);
+
+            bool is5xxError;
+            HttpResponse response;
+
+            try
+            {
+                response = await ExecuteWithCancellationTokenAsync(endpoint, headers, null, HttpMethod.Get, cts.Token).ConfigureAwait(false);
+
+                if (response.StatusCode == HttpStatusCode.OK)
+                {
+                    return response;
+                }
+
+                logger.Info(string.Format(CultureInfo.InvariantCulture,
+                    MsalErrorMessage.HttpRequestUnsuccessful,
+                    (int)response.StatusCode, response.StatusCode));
+
+
+                is5xxError = (int)response.StatusCode >= 500 && (int)response.StatusCode < 600;
+            }
+            catch (TaskCanceledException exception)
+            {
+                logger.Error("The HTTP request failed or it was canceled. " + exception.Message);
+                throw new MsalServiceException(
```

Comment:
It's better to let the cancellation exception propagate to the are of the code where you use it instead of catching and rethrowing a new exception. `MsalServiceExceptions` are signals for the user, not for your own internal logic.

Reply:
Updated